crates/mod-crypto/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+    pub value: String,
+    pub algorithm: String,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct LuaCfgDecrypt {
+    pub key: KeySource,
+    pub value: Vec<u8>,
+    pub algorithm: String,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+pub enum AesKey {
+    Aes128([u8; 16]),
+    Aes256([u8; 32]),
+}
+
+fn is_hex(vec: &[u8]) -> bool {
+    let hex_str = String::from_utf8_lossy(vec);
+    hex_str.chars().all(|c| c.is_digit(16)) && hex_str.len() % 2 == 0
+}
+
+impl AesKey {
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, String> {
+        let decoded_bytes = if is_hex(bytes) {
+            decode(bytes).map_err(|e| format!("Failed to decode hex: {}", e))?
+        } else {
+            bytes.to_vec()
+        };
+        match decoded_bytes.len() {
+            16 => Ok(AesKey::Aes128(decoded_bytes.try_into().unwrap())),
+            32 => Ok(AesKey::Aes256(decoded_bytes.try_into().unwrap())),
+            _ => Err("Key length must be 16 or 32 bytes".to_string()),
+        }
+    }
+}
+
+fn aes_encrypt_block(plaintext: &str, params: AesParams) -> Result<Vec<u8>, anyhow::Error> {
+    let mut buf_ciphertext = plaintext.as_bytes().to_vec();
+    let enc_key: PaddedBlockEncryptingKey;
+
+    match params.algorithm {
+        AesAlgo::Ecb() => {
+            enc_key = match params.key {
+                AesKey::Aes128(k) => {
+                    let key = UnboundCipherKey::new(&AES_128, &k)?;
+                    PaddedBlockEncryptingKey::ecb_pkcs7(key)?
+                }
+                AesKey::Aes256(k) => {
+                    let key = UnboundCipherKey::new(&AES_256, &k)?;
+                    PaddedBlockEncryptingKey::ecb_pkcs7(key)?
+                }
+            };
+            // don't return IV vector for ecb
+            enc_key.encrypt(&mut buf_ciphertext)?;
+            Ok(buf_ciphertext)
+        }
+        AesAlgo::Cbc() => {
+            enc_key = match params.key {
+                AesKey::Aes128(k) => {
+                    let key = UnboundCipherKey::new(&AES_128, &k)?;
+                    PaddedBlockEncryptingKey::cbc_pkcs7(key)?
+                }
+                AesKey::Aes256(k) => {
+                    let key = UnboundCipherKey::new(&AES_256, &k)?;
+                    PaddedBlockEncryptingKey::cbc_pkcs7(key)?
+                }
+            };
+
+            // context contain IV vector
+            let context = enc_key.encrypt(&mut buf_ciphertext)?;
+
+            match context {
+                DecryptionContext::Iv128(iv) => {
+                    let mut result = iv.as_ref().to_vec();
+                    result.extend_from_slice(&buf_ciphertext); // Append the actual ciphertext after the IV
+                    Ok(result)
+                }
+                _ => Err(anyhow!("Unexpected IV context for encrypting in CBC mode")),
+            }
+        }
+    }
+}
+
+
+struct DecryptResult(Vec<u8>);
+impl LuaUserData for DecryptResult {
+    fn add_fields<F: UserDataFields<Self>>(fields: &mut F) {
```

Comment:
I realize that this is a precedent set by the digest module; I think it is convenient to do this sort of thing, but we should probably factor it out so that it can be re-used.

I'd suggest that we rename the digest module's `DigestResult` to `BinaryResult` and make it public, then re-use it in this module in place of `DecryptResult` so that we only have a single implementation that can be consistently used throughout this and similar future crates.